replay/models/nmslib_hnsw.py

```diff
@@ -0,0 +1,646 @@
+                        num_threads=1,
+                    )
+                    pd_res = pd.DataFrame(
+                        neighbours, columns=["item_idx", "distance"]
+                    )
+
+                    return pd_res
+
+        cols = []
+        if index_type == "sparse":
+            cols += ["user_idx", "vector_items", "vector_relevances"]
+        else:
+            cols.append(features_col)
+        if filter_seen_items:
+            cols = cols + ["num_items", "seen_item_idxs"]
+
+        res = user_vectors.select(
+            "user_idx",
+            infer_index(*cols).alias("neighbours"),
+        )
+
+        res = self._unpack_infer_struct(res)
+
+        return res
+
+    def _save_nmslib_hnsw_index(self, path, sparse=False):
+        """Method save (copy) index from hdfs (or local) to `path` directory.
+        `path` can be an hdfs path or a local path.
+
+        Args:
+            path (_type_): directory where to dump (copy) the index
+        """
+
+        params = self._nmslib_hnsw_params
+
+        if params["build_index_on"] == "executor":
+            index_path = params["index_path"]
+        elif params["build_index_on"] == "driver":
+            index_path = SparkFiles.get(
+                f"{INDEX_FILENAME}_{self._spark_index_file_uid}"
+            )
+        else:
+            raise ValueError("Unknown 'build_index_on' param.")
+
+        source = get_filesystem(index_path)
+        target = get_filesystem(path)
+        self.logger.debug("Index file coping from '%s' to '%s'", index_path, path)
+
+        source_paths = []
+        target_paths = []
+        if sparse:
+            source_paths.append(source.path)
+            source_paths.append(source.path + ".dat")
+            index_file_target_path = os.path.join(target.path, INDEX_FILENAME)
+            target_paths.append(index_file_target_path)
+            target_paths.append(index_file_target_path + ".dat")
+        else:
+            source_paths.append(source.path)
+            index_file_target_path = os.path.join(target.path, INDEX_FILENAME)
+            target_paths.append(index_file_target_path)
+
+        if source.filesystem == FileSystem.HDFS:
+            source_filesystem = fs.HadoopFileSystem.from_uri(source.hdfs_uri)
+        else:
+            source_filesystem = fs.LocalFileSystem()
+        if target.filesystem == FileSystem.HDFS:
+            destination_filesystem = fs.HadoopFileSystem.from_uri(
+                target.hdfs_uri
+            )
+        else:
+            destination_filesystem = fs.LocalFileSystem()
+
+        for source_path, target_path in zip(source_paths, target_paths):
+            fs.copy_files(
+                source_path,
+                target_path,
+                source_filesystem=source_filesystem,
+                destination_filesystem=destination_filesystem,
+            )
+            # param use_threads=True (?)
+
+    def _load_nmslib_hnsw_index(self, path: str, sparse=False):
```

Comment:
please avoid blob files (putting a lot of functionality into one class), that could easily violates single responsibility, dependency inversion principles